google/cloud/spanner/row.h

```diff
@@ -0,0 +1,179 @@
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_ROW_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_ROW_H_
+
+#include "google/cloud/spanner/version.h"
+#include <cstdint>
+#include <string>
+#include <tuple>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+/**
+ * The `Row<Ts...>` class template represents a heterogeneous set of C++ values.
+ *
+ * The values stored in a row may have any of the valid Spanner types described
+ * in value.h. Each value in a row is identified by its column index, with the
+ * first value corresponding to column 0. The number of columns in the row is
+ * given by the `Row::size()` member function (This exactly matches the number
+ * of types the Row was created with).
+ *
+ * A `Row<Ts...>` is a regular C++ type, supporting default construction, copy,
+ * assignment, move, and equality as expected. It should typically be
+ * constructed with values using the non-member `spanner::MakeRow(Ts...)`
+ * function template (see below). Once an instance is created, you can access
+ * the values in the row using the `Row::get()` overloaded functions.
+ *
+ * Example:
+ *
+ *     spanner::Row<bool, std::int64_t, std::string> row;
+ *     row = spanner::MakeRow(true, 42, "hello");
+ *     static_assert(row.size() == 3, "Created with three types");
+ *
+ *     // Gets all values as a std::tuple
+ *     std::tuple<bool, std::int64_t, std::string> tup = row.get();
+ *
+ *     // Gets one value at the specified column index
+ *     std::int64_t i = row.get<1>();
+ *
+ *     // Gets the values from the specified columns (any order), and returns a
+ *     // tuple. Since C++17, this tuple will work with structured bindings to
+ *     // allow assigning to individually  named variables.
+ *     auto [name, age] = row.get<2, 1>();
+ *
+ */
+template <typename... Types>
+class Row {
+  template <std::size_t I>
+  using ColumnType = typename std::tuple_element<I, std::tuple<Types...>>::type;
+
+ public:
+  /// Regular value type, supporting copy, assign, move, etc.
+  Row() = default;
+  Row(Row const&) = default;
+  Row& operator=(Row const&) = default;
+  Row(Row&) = default;
+  Row& operator=(Row&&) = default;
+
+  /**
+   * Constructs a Row from the specified values.
+   *
+   * See also the `MakeRow()` helper function template for a convenient way to
+   * make rows without having to also specify the value's types.
+   */
+  template <typename... Ts>
+  explicit Row(Ts... ts) : values_(std::forward<Types>(ts)...) {}
+
+  /// Returns the number of columns in this row.
+  constexpr std::size_t size() const { return sizeof...(Types); }
+
+  /**
+   *  Returns the value at position `I`.
+   *
+   *  Example:
+   *
+   *      auto row = MakeRow(true, "foo");
+   *      assert(row.get<0>() == true);
+   *      assert(row.get<1>() == "foo");
+   */
+  template <std::size_t I>
+  ColumnType<I> get() const {
```

Comment:
Should we have a non-const overload like `tuple` does? What if the row has some large columns that I want to "move out", should we have a `&&` overload too?

Reply:
Good point. Done.